### DISCUSSION TITLE
Intermittent cluster-wide transaction failures with atomic guard enabled

### DISCUSSION THREAD

**Discussion by krivulcik** [ravendb/ravendb#14630] [Q&A]
The issue first described here: https://groups.google.com/g/ravendb/c/iMOqmY9LlNs might have another cause.

This is on a project where the client library is 5.3.102. The cluster runs 5.3.103.

Because of the issue https://github.com/ravendb/ravendb/discussions/14472#discussioncomment-3136537 , which was caused by disabling the atomic guards, we have enabled the atomic guards in our codebase (removed the setting `DisableAtomicDocumentWritesInClusterWideTransaction = true`).

The error has occurred again: 

```
Raven.Client.Exceptions.ClusterTransactionConcurrencyException: Failed to execute cluster transaction due to the following issues: Guard compare exchange value 'rvn-atomic/[DOC_ID]' index does not match the transaction index's 0 change vector on [DOC_ID]
Concurrency check failed for putting the key 'rvn-atomic/[DOC_ID]'. Requested index: 0, actual index: 18284036
   at Raven.Client.Exceptions.ExceptionDispatcher.ThrowConflict(ExceptionSchema schema, BlittableJsonReaderObject json) in C:\Builds\RavenDB-Stable-5.3\53016\src\Raven.Client\Exceptions\ExceptionDispatcher.cs:line 167
   at Raven.Client.Exceptions.ExceptionDispatcher.Throw(JsonOperationContext context, HttpResponseMessage response, Action`1 additionalErrorInfo) in C:\Builds\RavenDB-Stable-5.3\53016\src\Raven.Client\Exceptions\ExceptionDispatcher.cs:line 78
   at Raven.Client.Http.RequestExecutor.HandleUnsuccessfulResponse[TResult](ServerNode chosenNode, Nullable`1 nodeIndex, JsonOperationContext context, RavenCommand`1 command, HttpRequestMessage request, HttpResponseMessage response, String url, SessionInfo sessionInfo, Boolean shouldRetry, CancellationToken token) in C:\Builds\RavenDB-Stable-5.3\53016\src\Raven.Client\Http\RequestExecutor.cs:line 1425
   at Raven.Client.Http.RequestExecutor.ExecuteAsync[TResult](ServerNode chosenNode, Nullable`1 nodeIndex, JsonOperationContext context, RavenCommand`1 command, Boolean shouldRetry, SessionInfo sessionInfo, CancellationToken token) in C:\Builds\RavenDB-Stable-5.3\53016\src\Raven.Client\Http\RequestExecutor.cs:line 903
   at Raven.Client.Http.RequestExecutor.ExecuteAsync[TResult](ServerNode chosenNode, Nullable`1 nodeIndex, JsonOperationContext context, RavenCommand`1 command, Boolean shouldRetry, SessionInfo sessionInfo, CancellationToken token) in C:\Builds\RavenDB-Stable-5.3\53016\src\Raven.Client\Http\RequestExecutor.cs:line 903
   at Raven.Client.Documents.Session.AsyncDocumentSession.SaveChangesAsync(CancellationToken token) in C:\Builds\RavenDB-Stable-5.3\53016\src\Raven.Client\Documents\Session\AsyncDocumentSession.cs:line 169
   at [my code]
```

However, it is not deterministic: it only occurred once, even though this code was executed multiple times before and after the error.

The code where the issue occurred is very simple. Even though it is wrapped in our custom wrappers, it boils down to the following:

```c#
using (var session = store.OpenAsyncSession(new SessionOptions { TransactionMode = TransactionMode.ClusterWide }))
{
    var doc = await session.LoadAsync<Doc>("DOC_ID");

    doc.Finished = DateTime.UtcNow;
    doc.Result = "some value";

    await session.StoreAsync(doc);

    await session.SaveChangesAsync();
}
```

The whole thing is part of long-running background processing and is wrapped in a try-catch block. The catch block does the same thing, except it sets the fields `Failed` and `FailureReason` instead of `Finished` and `Result`. This succeeded, so the document contains the DateTime of the failure and the exception message.

The document has the following change vector: `RAFT:941-38ygWxiPokeef7gCSGHN6g TRXN:18284207-gUCijslFrEygc/HDrzmPMA` .

Since deploying the application with the atomic guard enabled this morning, this code was successfully executed 14 times before the error occurred and twice after the error.

It's definitely not one of the cases Igal described here: https://groups.google.com/g/ravendb/c/iMOqmY9LlNs/m/dmqqT7EGBQAJ (storing a new instance without using a loaded document, or deleting/recreating it).

It occurs only intermittently.

Can it be caused by the client-server version mismatch? I'll upgrade the cluster and client libraries to see if the issue persists. (This project was missed when we were upgrading to 5.3.103.)

Can it be caused by explicitly calling `session.Store[Async]`? Although not necessary for loaded/changed documents, we use it to express intent to store the document.

Could it possibly be caused by a topology change at an unfortunate moment? The document is created in one application (web), but all subsequent processing is done in another, separate, application. This application changes the document multiple times (first marks it as Started, and then as either Finished or Failed).

Any suggestions for further investigation are greatly appreciated.

**Comment by ayende** [comment]
It is probably an issue with concurrent transactions modifying the same document. 
Looks like you edited the message, so we can't tell exactly in what manner, but I'll bet that the document was deleted by another cluster wide transaction.

**Reply by krivulcik** [reply]
I'll be out of the office until next week, so I can't look at this very closely right away. I'll get back to it later.

By my editing the message, do you mean the exception message and stack trace? I redacted the ID of the document from the message, as it contained an email address. The `DOC_ID` was something like `req-se-user@example.com-2022-07-27-10-20-30`.

We don't have any code to delete documents from the collection in question, which makes the delete-change race condition unlikely. Also, the processing is done sequentially, so there *shouldn't* be anything else touching that document in any way at that time (I'll try to verify this, but the logs suggest that the processing application was working as usual).

**Comment by karmeli87** [comment]
From the exception `Requested index: 0, actual index: 18284036` it looks like you have a race between creating a _new_ document and modifying an existing one.

Each such cluster operation (in this case cluster-wide tx) creates a unique id which is send along the command, so if the command reach the server but the client still failover we will not execute the command twice, because we will check if we saw this unique id first.
But we use a cache of 2K items to store those unique id, so it might happened under very high number of cluster operation that we will drop the unique id of that command and when the client will failover it will consider it as a new command and will try to execute it, resulting in the exception above. 

But this is only a theory :)

**Reply by krivulcik** [reply]
I'll investigate more closely next week, but preliminarily, I don't think that this is the case.

- Would the new document and the existing document have to have the same ID? So the new document is trying to replace the existing document because of the ID? This is unlikely, as the ID contains a timestamp (second-precision) when it was created. This exception was thrown later, during processing (seconds later, but later), so even if we somehow managed to attempt to create a new document, it would have a different ID.
- We don't have a high number of cluster-wide transactions going on. Nowhere near 2k, like, ever, I'd say (yet).

I'll try to reproduce this behavior later. I'm hoping that it occurs with moderate testing/load, as this issue occurred pretty soon after deploying the changed code

**Reply by karmeli87** [reply]
- So if 2 docs created under 1 sec, they can get the same Id, no?
- The scenario above is about a failover of creating a new document (can happened for modifying as well, but with a different `Requested index` in the exception which is not 0).

It's not about only cluster-tx, but any cluster operation like update subscription state, identities, etc.. you can all of those commands and the 2k cache in this EP `/admin/debug/cluster/history-logs`

**Reply by krivulcik** [reply]
> if 2 docs created under 1 sec, they can get the same Id

Yes, they can, but this will happen only when the user double-submits the form or something like that, and we don't care in such cases. One will likely succeed and one will likely fail, which is fine in this particular use case. The ID contains the user ID, so different users won't get into conflict with each other.

Also, this particular exception occurred during an update where the document is loaded, changed, and immediately stored in a separate, isolated async session.

I'll try to induce this issue next week.

**Comment by krivulcik** [comment]
I might have a clue about this issue. While we are not deleting any documents anywhere, I found a situation in our code where the atomic guard was sometimes enabled and sometimes disabled.

This leads to an interesting situation in the studio where both "modified document" and "deleted document" warnings are shown:

![image](https://user-images.githubusercontent.com/7997055/185375380-233664aa-a10b-4ea9-bd3e-fc863503b7f4.png)

Consider the test below (I used a single-node transaction to create the document):
 - At first, the document is created with a CV like `A:1-K9CQvO89hEyciOoEIDNnYQ` (single-node transaction).
 - Then, it is changed and gets a CV like `RAFT:1-scPSb9839UyNP+wKAl16ng` (cluster-wide transaction with atomic guard disabled).
 - Then, it is changed and gets a CV like `RAFT:2-scPSb9839UyNP+wKAl16ng,TRXN:4-RHdmLAYoMkKD8yxQoOE3VA` (cluster-wide transaction with atomic guard enabled).
 - At last, it is changed and gets a CV like `RAFT:3-scPSb9839UyNP+wKAl16ng` (same as the first change above - cluster-wide transaction with atomic guard disabled).

Could one of the changes in the sequence lead to the issue above? It seems that RavenDB detects the document deletion in *some* way, so the situation would make sense.

Would it be possible to handle it on the database level? We will fix this kind of inconsistent approach (it was a mismatch in configuration between sync and async sessions), but even though the usage might not be valid, leading to similar elusive issues isn't great.

Test code:

```C#
using Raven.Client.Documents.Session;
using Raven.TestDriver;
using Xunit;

namespace RavenDB5Tests.Tests
{
    public class ClusterWideTransactions : RavenTestDriver
    {
        [Fact]
        public void ClusterWideTransactionsTest()
        {
            ConfigureServer(TestServerOptions.UseFiddler());

            using var store = GetDocumentStore();
            WaitForUserToContinueTheTest(store);

            using (var session = store.OpenSession(new SessionOptions { TransactionMode = TransactionMode.SingleNode }))
            {
                var doc = new Doc { Id = "doc-1", NumVal = 1 };
                session.Store(doc);
                session.SaveChanges();
            }

            using (var session = store.OpenSession(new SessionOptions { TransactionMode = TransactionMode.ClusterWide, DisableAtomicDocumentWritesInClusterWideTransaction = true }))
            {
                var doc = session.Load<Doc>("doc-1");
                doc.NumVal++;
                session.Store(doc);
                session.SaveChanges();
            }

            using (var session = store.OpenSession(new SessionOptions { TransactionMode = TransactionMode.ClusterWide, DisableAtomicDocumentWritesInClusterWideTransaction = false }))
            {
                var doc = session.Load<Doc>("doc-1");
                doc.NumVal++;
                session.Store(doc);
                session.SaveChanges();
            }

            using (var session = store.OpenSession(new SessionOptions { TransactionMode = TransactionMode.ClusterWide, DisableAtomicDocumentWritesInClusterWideTransaction = true }))
            {
                var doc = session.Load<Doc>("doc-1");
                doc.NumVal++;
                session.Store(doc);
                session.SaveChanges();
            }
        }

        public class Doc
        {
            public string Id { get; set; }
            public int NumVal { get; set; }
        }
    }
}
```

**Reply by ayende** [reply]
You are correct that this will cause issues. I think that the delete notice is not relevant for our purposes, 

Basically, when you mix cluster wide transactions with and without atomic guard, you mess up expectations.

**Reply by ayende** [reply]
@karmeli87 can you make a unit test and figure out what sort of error we need to throw here?

**Reply by karmeli87** [reply]
Thanks @krivulcik for digging into it. I think mixing cluster-wide & non-cluster wide is expected.
What is not expected here is to mix 2 different type of cluster-wide transactions.
I've create an issue that you can follow https://issues.hibernatingrhinos.com/issue/RavenDB-19177.
